apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIrBuilder.kt

```diff
@@ -1,205 +1,298 @@
+        description = "",
+        name = "data",
+        arguments = emptyList(),
+        directives = emptyList(),
+        type = GQLNamedType(name = rootTypeDefinition.name)
+    )
+    val dataField = GQLField(
+        name = "data",
+        alias = null,
+        sourceLocation = com.apollographql.apollo.compiler.frontend.gql.SourceLocation.UNKNOWN,
+        arguments = null,
+        directives = emptyList(),
+        selectionSet = selectionSet
     ).buildBackendIrField(
         selectionKey = selectionKey,
         generateFragmentImplementations = true,
+        dataFieldDefinition
     )
-    val variables = this.variables.map { variable ->
+    val variables = this.variableDefinitions.map { variable ->
       BackendIr.Variable(
           name = variable.name,
-          type = schema.resolveType(variable.type)
+          type = variable.type.toSchemaType(schema)
       )
     }
+    val fragmentNames = usedFragmentNames(schema, fragmentDefinitions)
     return BackendIr.Operation(
         name = normalizedName,
-        operationName = this.operationName,
-        targetPackageName = this.packageName,
-        operationType = typeRef,
-        comment = this.description,
+        operationName = name!!,
+        targetPackageName = packageNameProvider?.operationPackageName(filePath = sourceLocation.filePath ?: "") ?: "",
+        operationType = IntrospectionSchema.TypeRef(
+            IntrospectionSchema.Kind.OBJECT,
+            rootTypeDefinition.name
+        ),
+        comment = this.description ?: "",
         variables = variables,
-        definition = this.sourceWithFragments,
+        definition = (toUtf8WithIndents() + "\n" + fragmentNames.map { fragmentDefinitions[it]!!.toUtf8WithIndents() }.joinToString("\n")).trimEnd('\n'),
         dataField = dataField,
     )
   }
 
-  private val Operation.operationTypeRef: IntrospectionSchema.TypeRef
-    get() {
-      return when {
-        this.isQuery() -> schema.resolveType(schema.queryType)
-        this.isMutation() -> checkNotNull(schema.resolveType(schema.mutationType ?: "")) {
-          "Can't resolve GraphQL mutation operation type"
-        }
-        this.isSubscription() -> checkNotNull(schema.resolveType(schema.subscriptionType ?: "")) {
-          "Can't resolve GraphQL mutation operation type"
-        }
-        else -> throw IllegalStateException("Unsupported GraphQL operation type: $operationType")
-      }
-    }
-
-  private fun Operation.normalizeOperationName(): String {
+  private fun GQLOperationDefinition.normalizeOperationName(): String {
     fun normalizeOperationName(
         useSemanticNaming: Boolean,
         operationNameSuffix: String,
     ): String {
-      return if (useSemanticNaming && !operationName.endsWith(operationNameSuffix)) {
-        operationName.capitalize() + operationNameSuffix
+      require(name != null) {
+        "anonymous operations are not supported. This should have been caught during validation"
```

Comment:
if this should be caught by validation, do we need check it here again? If yes should we update this message to be user friendly?

Reply:
I added the check mainly so that the compiler can infer the non-nullability and not force me to type `name!!` later on but really this should never happen™️. I'm happy to use `name!!` or anything else.